firedrake/mesh.py

```diff
@@ -2366,6 +2381,57 @@ def _pic_swarm_in_plex(plex, coords, fields=[]):
     return swarm
 
 
+def _parent_mesh_embedding(vertex_coords, parent_mesh):
+    """Find the parent cells and local coords for vertices on this rank."""
+    if len(vertex_coords) > 0:
+        vertex_coords = _on_rank_vertices(vertex_coords, parent_mesh)
+
+    num_vertices = len(vertex_coords)
+    max_num_vertices = parent_mesh.comm.allreduce(num_vertices, op=MPI.MAX)
+
+    gdim = parent_mesh.geometric_dimension()
+    tdim = parent_mesh.topological_dimension()
+
+    parent_cell_nums = np.empty(num_vertices, dtype=IntType)
+    reference_coords = np.empty((num_vertices, tdim), dtype=RealType)
+    valid = np.full(num_vertices, False)
+
+    # Create an out of mesh point to use in locate_cell when needed
+    out_of_mesh_point = np.full((1, gdim), np.inf)
+
+    for i in range(max_num_vertices):
+        if i < num_vertices:
+            parent_cell_num, reference_coord = \
+                parent_mesh.locate_cell_and_reference_coordinate(vertex_coords[i])
+            if parent_cell_num is not None and parent_cell_num < parent_mesh.cell_set.size:
```

Comment:
Comment here please about `parent_cell_num < parent_mesh.cell_set.size` avoiding putting vertices in the parent mesh halo